include/pybind11/numpy.h

```diff
@@ -444,31 +444,30 @@ class array : public buffer {
 
     /// Pointer to the contained data. If index is not provided, points to the
     /// beginning of the buffer. May throw if the index would lead to out of bounds access.
-    template<typename... Ix> const void* data(Ix&&... index) const {
+    template<typename... Ix> const void* data(Ix... index) const {
         return static_cast<const void *>(PyArray_GET_(m_ptr, data) + offset_at(index...));
     }
 
     /// Mutable pointer to the contained data. If index is not provided, points to the
     /// beginning of the buffer. May throw if the index would lead to out of bounds access.
     /// May throw if the array is not writeable.
-    template<typename... Ix> void* mutable_data(Ix&&... index) {
+    template<typename... Ix> void* mutable_data(Ix... index) {
         check_writeable();
         return static_cast<void *>(PyArray_GET_(m_ptr, data) + offset_at(index...));
     }
 
     /// Byte offset from beginning of the array to a given index (full or partial).
     /// May throw if the index would lead to out of bounds access.
-    template<typename... Ix> size_t offset_at(Ix&&... index) const {
-        if (sizeof...(index) > ndim())
-            fail_dim_check(sizeof...(index), "too many indices for an array");
-        return get_byte_offset(index...);
+    template<typename... Ix> size_t offset_at(Ix... index) const {
+        check_dimensions(index...);
```

Comment:
This seems wrong. Note that the check here is for `>` whereas in other places it is for `!=`, otherwise it would have been implemented like you suggest here via a `check_dimensions(...)` helper.